crates/ide/src/syntax_highlighting/inject.rs

```diff
@@ -101,12 +163,33 @@ pub(super) fn doc_comment(hl: &mut Highlights, node: &SyntaxNode) {
     // Replace the original, line-spanning comment ranges by new, only comment-prefix
     // spanning comment ranges.
     let mut new_comments = Vec::new();
-    for comment in doc_comments {
-        match comment.text().find(RUSTDOC_FENCE) {
+    let mut string;
+    for attr in attributes.by_key("doc").attrs() {
+        let src = attr.to_src(&owner);
```

Comment:
This `to_src` call actually has O(n^2) runtime as it recollects all doc items every call then takes the nth item from there, should probably look into a better mapping

Reply:
Yeah, I'm pretty unsure what would make a good API for attributes. Let me know if you have ideas.

Reply:
Yeah, most of our `.source` calls are quadratic. The fundamental thing here is that we can't ask for sources of a *single* thing. Rather, we can ask the parent "what are the sources of all your children?", and there needs to be some caching layer there. `impl HasSource for Variant` is also quadratic, for example, and that's not really right. 

Ideally, this caching should be done transparently by salsa, but this is the bit where we need gc unfortunately.